File: experiments/05-deployed-do-memory-brain/harness.mjs
// Experiment 05 harness — runs the same four contract tests as experiments
// 01 and 03 (same-client R-A-W, second-client visibility, concurrent writes,
// latency) but is shaped to also be pointed at a *deployed* workers.dev URL.
//
// The harness itself never deploys anything. It speaks plain HTTP and uses
// Node 22 native `fetch`. Two distinct fetch invocations model two distinct
// clients — there is no shared in-process state between them, so cross-client
// visibility is a real property of the DO, not of the harness.
//
// Flags:
//   --base <url>       endpoint to hit. local: http://127.0.0.1:8875
//                      remote: https://dejavu-exp05-...workers.dev
//   --out <file>       output markdown path (default RESULT.md)
//   --mode local|remote   marks the result for clarity (default local)
//   --label <string>      free-form label for the RESULT (e.g. "deployed-iad")
//   --reset            POST /reset before testing. Only succeeds if the
//                      worker was deployed with DEJAVU_EXP_ALLOW_RESET=1; on
//                      a remote URL this is the only safe way to start
//                      from a clean DO.
//   --concurrent <n>   concurrent writers in test 3 (default 64)
//   --sequential <n>   sequential ops in test 4 (default 50)
//   --warm <n>         warm-up requests before timing (default 3, helpful for
//                      remote runs where the first request pays DO cold-start)

import { writeFileSync } from "node:fs";
import { argv } from "node:process";

const args = parseArgs(argv.slice(2));
const base = (args.base ?? "http://127.0.0.1:8875").replace(/\/+$/, "");
const out = args.out ?? "RESULT.md";
const mode = args.mode ?? "local";
const label = args.label ?? "";
const doReset = !!args.reset;
const N = clampInt(args.concurrent, 64, 1, 512);
const M = clampInt(args.sequential, 50, 1, 1000);
const warm = clampInt(args.warm, 3, 0, 50);

function parseArgs(arr) {
  const o = {};
  for (let i = 0; i < arr.length; i++) {
    const a = arr[i];
    if (a.startsWith("--")) {
      const key = a.slice(2);
      const next = arr[i + 1];
      if (next === undefined || next.startsWith("--")) {
        o[key] = true; // boolean flag
      } else {
        o[key] = next;
        i++;
      }
    }
  }
  return o;
}

function clampInt(raw, def, min, max) {
  if (raw == null || raw === true) return def;
  const n = Number.parseInt(raw, 10);
  if (!Number.isFinite(n)) return def;
  return Math.max(min, Math.min(max, n));
}

async function jget(path) {
  const r = await fetch(base + path);
  return { status: r.status, body: await r.json() };
}
async function jpost(path, body) {
  const r = await fetch(base + path, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify(body ?? {}),
  });
  return { status: r.status, body: await r.json() };
}

function pct(arr, p) {
  if (arr.length === 0) return 0;
  const sorted = [...arr].sort((a, b) => a - b);
  const idx = Math.min(sorted.length - 1, Math.floor((p / 100) * sorted.length));
  return sorted[idx];
}
function mean(arr) {
  return arr.reduce((a, b) => a + b, 0) / Math.max(1, arr.length);
}
const fmt = (n) => (Math.round(n * 100) / 100).toFixed(2);

async function main() {
  // ---- preflight: refuse to talk to anything that isn't /health-shaped ----
  const health = await jget("/health");
  if (health.status !== 200 || !health.body || health.body.ok !== true) {
    console.error(
      `preflight: GET ${base}/health did not return ok=true (status=${health.status}). ` +
        `Refusing to run — pointed at the wrong URL?`,
    );
    process.exit(1);
  }
  const ftsMode = health.body.fts;
  const ftsError = health.body.ftsError;
  const build = health.body.build ?? "unknown";
  const resetEnabled = !!health.body.resetEnabled;
  const lastIdAtStart = health.body.lastId ?? 0;

  // ---- optional reset ----
  let resetReport = "not requested";
  if (doReset) {
    const rs = await jpost("/reset", {});
    if (rs.status === 200 && rs.body?.ok) {
      resetReport = `cleared ${rs.body.cleared} rows`;
    } else {
      resetReport = `reset refused (status=${rs.status}, body=${JSON.stringify(rs.body)})`;
      console.warn(`[harness] /reset returned ${rs.status}; continuing without a clean DO`);
    }
  }

  // Re-probe health after reset so the rest of the run anchors off the post-reset baseline.
  const health2 = await jget("/health");
  const baseLastId = health2.body?.lastId ?? lastIdAtStart;

  // ---- warm-up (don't time, but exercise the path) ----
  for (let i = 0; i < warm; i++) {
    const tag = `warm-${Math.random().toString(36).slice(2, 8)}`;
    await jpost("/remember", { client: "warm", text: `warmup ${tag}` });
    await jget(`/recall?q=${encodeURIComponent(tag)}`);
  }
  // After warm-up the lastId has advanced; refresh the baseline so concurrency
  // assertions in test 3 still hold.
  const healthAfterWarm = await jget("/health");
  const baseAfterWarm = healthAfterWarm.body?.lastId ?? baseLastId;

  // ---- Test 1: same-client read-after-write ----
  const tag1 = `alpha-${Math.random().toString(36).slice(2, 10)}`;
  const w1 = await jpost("/remember", { client: "alice", text: `hello ${tag1} world` });
  const r1 = await jget(`/recall?q=${encodeURIComponent(tag1)}`);
  const test1 = {
    id: w1.body.id,
    ulid: w1.body.ulid,
    durable: w1.body.durable,
    sawOwn: r1.body.hits.some((h) => h.ulid === w1.body.ulid),
    hitCount: r1.body.hits.length,
  };

  // ---- Test 2: cross-client visibility ----
  const tag2 = `beta-${Math.random().toString(36).slice(2, 10)}`;
  const w2 = await jpost("/remember", { client: "alice", text: `cross ${tag2} client` });
  const r2 = await jget(`/recall?q=${encodeURIComponent(tag2)}`);
  const test2 = {
    id: w2.body.id,
    ulid: w2.body.ulid,
    sawAlice: r2.body.hits.some((h) => h.ulid === w2.body.ulid),
    hitCount: r2.body.hits.length,
  };

  // ---- Test 3: concurrent writers ----
  const baseId = w2.body.id; // last known id before the burst
  const tag3 = `gamma-${Math.random().toString(36).slice(2, 10)}`;
  const t0 = performance.now();
  const writeLatencies = [];
  const results = await Promise.all(
    Array.from({ length: N }, (_, i) =>
      (async () => {
        const ts0 = performance.now();
        const r = await jpost("/remember", {
          client: `worker-${i}`,
          text: `${tag3} payload ${i}`,
        });
        writeLatencies.push(performance.now() - ts0);
        return r.body;
      })(),
    ),
  );
  const tBurst = performance.now() - t0;

  const ids = results.map((r) => r.id).filter((x) => typeof x === "number");
  const ulids = results.map((r) => r.ulid);
  const uniqUlids = new Set(ulids).size === ulids.length;
  const uniqIds = new Set(ids).size === ids.length;
  ids.sort((a, b) => a - b);
  const minId = ids[0];
  const maxId = ids[ids.length - 1];
  const contiguous = ids.every((v, i) => i === 0 || v === ids[i - 1] + 1);
  const startsAtBasePlus1 = minId === baseId + 1;

  const recallAll = await jget(`/recall?q=${encodeURIComponent(tag3)}&limit=200`);
  const recalledTexts = new Set(recallAll.body.hits.map((h) => h.text));
  const expected = Array.from({ length: N }, (_, i) => `${tag3} payload ${i}`);
  const missing = expected.filter((t) => !recalledTexts.has(t));

  const test3 = {
    N,
    errors: results.filter((r) => !r.ok).length,
    uniqUlids,
    uniqIds,
    contiguous,
    startsAtBasePlus1,
    idRange: [minId, maxId],
    missing: missing.length,
    burstMs: tBurst,
    writeP50: pct(writeLatencies, 50),
    writeP95: pct(writeLatencies, 95),
    writeP99: pct(writeLatencies, 99),
    writeMax: Math.max(...writeLatencies),
  };

  // ---- Test 4: light-load latency, sequential ----
  const rem = [];
  const rec = [];
  const e2e = [];
  let lostInLatency = 0;
  for (let i = 0; i < M; i++) {
    const tag = `delta-${i}-${Math.random().toString(36).slice(2, 8)}`;
    const a = performance.now();
    const wr = await jpost("/remember", { client: "lat", text: `lat ${tag}` });
    const b = performance.now();
    const rc = await jget(`/recall?q=${encodeURIComponent(tag)}`);
    const c = performance.now();
    rem.push(b - a);
    rec.push(c - b);
    e2e.push(c - a);
    if (!rc.body.hits.some((h) => h.ulid === wr.body.ulid)) {
      lostInLatency += 1;
      console.error(`latency loop: lost write ${wr.body.ulid}`);
    }
  }
  const test4 = {
    remember: { mean: mean(rem), p50: pct(rem, 50), p95: pct(rem, 95), p99: pct(rem, 99) },
    recall: { mean: mean(rec), p50: pct(rec, 50), p95: pct(rec, 95), p99: pct(rec, 99) },
    e2e: { mean: mean(e2e), p50: pct(e2e, 50), p95: pct(e2e, 95), p99: pct(e2e, 99) },
    lost: lostInLatency,
  };

  const stamp = new Date().toISOString();
  const isHttps = base.startsWith("https://");
  const headlineMode = mode === "remote" ? "deployed" : "local `wrangler dev`";

  const md = `# Experiment 05 — RESULT (${mode}${label ? `, ${label}` : ""})

- run timestamp: ${stamp}
- brain endpoint: \`${base}\`  (${isHttps ? "https" : "http"})
- prototype: real Cloudflare Worker + Durable Object — ${headlineMode}, DO uses \`ctx.storage.sql\`.
- worker build tag: \`${build}\`
- FTS mode: **${ftsMode}**${ftsError ? `  (FTS5 init error: \`${ftsError.replace(/`/g, "'")}\`)` : ""}
- reset endpoint at run start: ${resetEnabled ? "enabled" : "disabled"}
- /reset action: ${resetReport}
- lastId observed at start (after optional reset): ${baseLastId}
- lastId observed after warm-up (${warm} req): ${baseAfterWarm}

## Test 1 — same-client read-after-write

- receipt id: \`${test1.id}\`  ulid: \`${test1.ulid}\`  durable: \`${test1.durable}\`
- recalled own write: **${test1.sawOwn}**  (${test1.hitCount} hits)

## Test 2 — receipt → *different* client recall

- receipt id: \`${test2.id}\`  ulid: \`${test2.ulid}\`
- second client saw the write: **${test2.sawAlice}**  (${test2.hitCount} hits)

## Test 3 — concurrency (N=${test3.N})

- writers: ${test3.N}  errors: ${test3.errors}
- unique ulids: ${test3.uniqUlids}  unique ids: ${test3.uniqIds}
- contiguous id range: ${test3.contiguous}  starts at base+1: ${test3.startsAtBasePlus1}
- id range: [${test3.idRange[0]}, ${test3.idRange[1]}]
- texts missing from recall: ${test3.missing}
- write latency under load: p50=${fmt(test3.writeP50)}ms  p95=${fmt(test3.writeP95)}ms  p99=${fmt(test3.writeP99)}ms  max=${fmt(test3.writeMax)}ms
- burst wall-time: ${fmt(test3.burstMs)}ms

## Test 4 — latency (light load, sequential, M=${M})

- **remember**: mean=${fmt(test4.remember.mean)}ms  p50=${fmt(test4.remember.p50)}ms  p95=${fmt(test4.remember.p95)}ms  p99=${fmt(test4.remember.p99)}ms
- **recall**:   mean=${fmt(test4.recall.mean)}ms  p50=${fmt(test4.recall.p50)}ms  p95=${fmt(test4.recall.p95)}ms  p99=${fmt(test4.recall.p99)}ms
- **e2e**:      mean=${fmt(test4.e2e.mean)}ms  p50=${fmt(test4.e2e.p50)}ms  p95=${fmt(test4.e2e.p95)}ms  p99=${fmt(test4.e2e.p99)}ms
- read-after-write losses in light-load loop: ${test4.lost} (must be 0)

## Interpretation

The DO is the only authority and DO execution is single-threaded, so every
\`remember\` \`INSERT ... RETURNING id\` commits before the response leaves the
DO. A subsequent \`recall\` — whether issued by the same client (test 1) or
by an unrelated fetch caller (test 2) — hits the same DO and sees the row.
There is no replication lag because there is exactly one place memory lives.

Concurrency holds for the same reason: ${test3.N} parallel POSTs to
\`/remember\` are serialized by the DO's input gate and committed one at a time
by SQLite. Ids come back contiguous and ulids unique, with no missing payloads
on recall.

${
    mode === "remote"
      ? "Latency numbers above include real network RTT from the harness host to Cloudflare's edge plus DO placement + DO storage commit. They are a real measurement of how far reality is from local `wrangler dev`."
      : "Latency numbers above are local `wrangler dev` — workerd in-process. Useful as a floor; real CF numbers come from a `--mode remote` run against a deployed URL."
  }

FTS5 status: **${ftsMode}**. ${
    ftsMode === "fts5"
      ? "DO SQL accepted `CREATE VIRTUAL TABLE ... USING fts5(...)` and answered a MATCH query."
      : `DO SQL rejected FTS5 at init time:\n\n\u0060\u0060\u0060\n${ftsError ?? "(no message)"}\n\u0060\u0060\u0060\n\nThe harness fell back to \`LIKE\`-based recall in the same code path; visibility and concurrency claims still hold.`
  }

## Honest limitations of this prototype

- One DO id only (\`idFromName("singleton")\`). Sharding / placement / cold-start across many DOs are not modeled.
- Test 3's "burst" is bounded by the harness host's outbound HTTP concurrency, not by the DO itself.
- ${mode === "remote" ? "Remote latency includes everything from harness → CF edge → DO. A bad result can be CF placement, harness network, or DO — this experiment does not separate them." : "Local `wrangler dev` is not a Cloudflare durability proof. Numbers will differ on real CF; the *contract* should not."}
- Ulid is \`crypto.getRandomValues\`-based, not time-sortable. Matches the rest of the series; not production-grade.
- No auth, no quotas, no eviction, no embeddings.
`;

  writeFileSync(out, md);
  console.log(`wrote ${out}`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: experiments/01-do-memory-brain/harness.py
#!/usr/bin/env python3
"""Harness for experiment 01: DO-like memory brain.

Tests, in order:

1. Single-client remember -> immediate recall.
2. Receipt-then-second-client recall (a *separate* HTTP client object,
   modeling a different agent process talking to the same DO).
3. Concurrency: N parallel writers; verify (a) every receipt is unique, (b)
   ids form a contiguous monotonic range, (c) every text is recallable.
4. Latency: percentiles for remember and recall under light load.

Writes RESULT.md with observed numbers and notes.
"""
from __future__ import annotations

import argparse
import concurrent.futures
import json
import statistics
import time
import urllib.error
import urllib.request
from pathlib import Path


def _request(method: str, url: str, body: dict | None = None, timeout: float = 5.0) -> tuple[int, dict]:
    data = None
    headers = {"accept": "application/json"}
    if body is not None:
        data = json.dumps(body).encode("utf-8")
        headers["content-type"] = "application/json"
    req = urllib.request.Request(url, data=data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req, timeout=timeout) as r:
            payload = json.loads(r.read().decode("utf-8") or "{}")
            return r.status, payload
    except urllib.error.HTTPError as e:
        try:
            payload = json.loads(e.read().decode("utf-8") or "{}")
        except Exception:
            payload = {}
        return e.code, payload


class Client:
    """A standalone HTTP client. Each instance models a different agent."""

    def __init__(self, base: str, name: str):
        self.base = base.rstrip("/")
        self.name = name

    def remember(self, text: str, tags: list[str] | None = None) -> tuple[int, dict, float]:
        t0 = time.perf_counter()
        status, body = _request("POST", f"{self.base}/remember", {"text": text, "tags": tags or [], "author": self.name})
        return status, body, (time.perf_counter() - t0) * 1000.0

    def recall(self, q: str, limit: int = 20) -> tuple[int, dict, float]:
        t0 = time.perf_counter()
        status, body = _request("GET", f"{self.base}/recall?q={urllib.request.quote(q)}&limit={limit}")
        return status, body, (time.perf_counter() - t0) * 1000.0

    def stats(self) -> dict:
        _, body = _request("GET", f"{self.base}/stats")
        return body


def pct(values: list[float], p: float) -> float:
    if not values:
        return float("nan")
    s = sorted(values)
    k = (len(s) - 1) * p
    f = int(k)
    c = min(f + 1, len(s) - 1)
    return s[f] + (s[c] - s[f]) * (k - f)


def fmt_ms(x: float) -> str:
    return f"{x:.2f}ms"


def run(base: str, out_path: Path) -> None:
    alice = Client(base, "alice")
    bob = Client(base, "bob")

    results: dict = {"base": base}

    # ---- Test 1: same-client read-after-write ----
    text1 = f"read-after-write probe {time.time_ns()}"
    s, body, _ = alice.remember(text1, tags=["probe"])
    assert s == 201 and body.get("ok"), f"alice remember failed: {s} {body}"
    receipt1 = body["receipt"]
    s, body, _ = alice.recall(text1)
    assert s == 200 and body.get("ok"), f"alice recall failed: {s} {body}"
    hits1 = body["hits"]
    saw_self = any(h["ulid"] == receipt1["ulid"] for h in hits1)
    results["t1_same_client_raw"] = {
        "receipt": receipt1,
        "hitCount": len(hits1),
        "sawOwnWrite": saw_self,
    }
    assert saw_self, "same-client read-after-write FAILED"

    # ---- Test 2: receipt -> different client recall ----
    text2 = f"cross-client probe {time.time_ns()}"
    s, body, _ = alice.remember(text2, tags=["probe", "cross"])
    assert s == 201 and body.get("ok")
    receipt2 = body["receipt"]
    # bob is a fully separate Client instance: separate connection, separate
    # state. As soon as alice's POST returns, bob should see the row.
    s, body, _ = bob.recall(text2)
    assert s == 200 and body.get("ok")
    hits2 = body["hits"]
    bob_sees = any(h["ulid"] == receipt2["ulid"] for h in hits2)
    results["t2_cross_client"] = {
        "receipt": receipt2,
        "hitCount": len(hits2),
        "secondClientSaw": bob_sees,
    }
    assert bob_sees, "second-client read-after-receipt FAILED"

    # ---- Test 3: concurrency ----
    N = 64
    base_id = alice.stats()["latestId"]
    payloads = [f"concurrent probe {i} {uuid_short()}" for i in range(N)]
    writers = [Client(base, f"writer-{i % 4}") for i in range(N)]

    write_times: list[float] = []
    receipts: list[dict] = []
    errors: list[str] = []

    def do_write(i: int) -> None:
        try:
            s, body, dt = writers[i].remember(payloads[i], tags=["concurrent"])
            if s != 201 or not body.get("ok"):
                errors.append(f"#{i}: {s} {body}")
                return
            write_times.append(dt)
            receipts.append(body["receipt"])
        except Exception as exc:  # noqa: BLE001
            errors.append(f"#{i}: {exc!r}")

    with concurrent.futures.ThreadPoolExecutor(max_workers=16) as ex:
        list(ex.map(do_write, range(N)))

    ids = sorted(r["id"] for r in receipts)
    ulids = {r["ulid"] for r in receipts}
    # Verify each text is now recallable from a fresh reader (charlie).
    charlie = Client(base, "charlie")
    missing: list[int] = []
    for i, p in enumerate(payloads):
        _, body, _ = charlie.recall(p, limit=5)
        if not any(h["ulid"] == receipts[i]["ulid"] for h in body.get("hits", []) if "ulid" in h):
            # Fallback: if order of writes != order of receipts (because of
            # ThreadPoolExecutor.map result ordering), do a text-match check.
            found = any(h.get("text") == p for h in body.get("hits", []))
            if not found:
                missing.append(i)
    contiguous = ids == list(range(min(ids), min(ids) + N)) if ids else False
    monotonic_after_base = bool(ids) and min(ids) == base_id + 1
    results["t3_concurrency"] = {
        "writers": N,
        "errors": errors,
        "uniqueUlids": len(ulids) == N,
        "uniqueIds": len(set(ids)) == N,
        "contiguousIds": contiguous,
        "startsAtBasePlusOne": monotonic_after_base,
        "idRange": [min(ids), max(ids)] if ids else None,
        "missingFromRecall": missing,
        "writeMs": {
            "p50": pct(write_times, 0.50),
            "p95": pct(write_times, 0.95),
            "p99": pct(write_times, 0.99),
            "max": max(write_times) if write_times else float("nan"),
        },
    }
    assert not errors, f"concurrency errors: {errors}"
    assert len(ulids) == N, "ulid collisions"
    assert contiguous, f"ids not contiguous: {ids[:5]}..{ids[-5:]}"
    assert not missing, f"texts missing from recall: {missing}"

    # ---- Test 4: latency ----
    M = 100
    rem_ms: list[float] = []
    rec_ms: list[float] = []
    e2e_ms: list[float] = []
    for i in range(M):
        text = f"latency probe {i} {uuid_short()}"
        s, body, dt = alice.remember(text, tags=["latency"])
        assert s == 201 and body.get("ok")
        rem_ms.append(dt)
        s, body, dr = alice.recall(text, limit=5)
        assert s == 200 and body.get("ok")
        rec_ms.append(dr)
        e2e_ms.append(dt + dr)
    results["t4_latency"] = {
        "samples": M,
        "remember": {"p50": pct(rem_ms, .5), "p95": pct(rem_ms, .95), "p99": pct(rem_ms, .99), "mean": statistics.fmean(rem_ms)},
        "recall":   {"p50": pct(rec_ms, .5), "p95": pct(rec_ms, .95), "p99": pct(rec_ms, .99), "mean": statistics.fmean(rec_ms)},
        "e2e":      {"p50": pct(e2e_ms, .5), "p95": pct(e2e_ms, .95), "p99": pct(e2e_ms, .99), "mean": statistics.fmean(e2e_ms)},
    }

    # ---- Write RESULT.md ----
    out_path.write_text(render_result(results), encoding="utf-8")
    # Echo a compact JSON line for run.sh convenience.
    print(json.dumps({"summary": "ok", "tests": list(results.keys())}))


def uuid_short() -> str:
    import uuid
    return uuid.uuid4().hex[:8]


def render_result(r: dict) -> str:
    t1 = r["t1_same_client_raw"]
    t2 = r["t2_cross_client"]
    t3 = r["t3_concurrency"]
    t4 = r["t4_latency"]
    lines = []
    lines.append("# Experiment 01 — RESULT")
    lines.append("")
    lines.append(f"- brain endpoint: `{r['base']}`")
    lines.append("- prototype: single Python process, ThreadingHTTPServer, SQLite WAL, "
                 "global write lock around `BEGIN IMMEDIATE` (DO-shaped authority).")
    lines.append("")
    lines.append("## Test 1 — same-client read-after-write")
    lines.append("")
    lines.append(f"- receipt id: `{t1['receipt']['id']}`  ulid: `{t1['receipt']['ulid']}`  durable: `{t1['receipt']['durable']}`")
    lines.append(f"- recalled own write: **{t1['sawOwnWrite']}**  ({t1['hitCount']} hits)")
    lines.append("")
    lines.append("## Test 2 — receipt → *different* client recall")
    lines.append("")
    lines.append(f"- receipt id: `{t2['receipt']['id']}`  ulid: `{t2['receipt']['ulid']}`")
    lines.append(f"- second client saw the write: **{t2['secondClientSaw']}**  ({t2['hitCount']} hits)")
    lines.append("")
    lines.append("## Test 3 — concurrency")
    lines.append("")
    lines.append(f"- writers: {t3['writers']}  errors: {len(t3['errors'])}")
    lines.append(f"- unique ulids: {t3['uniqueUlids']}  unique ids: {t3['uniqueIds']}")
    lines.append(f"- contiguous id range: {t3['contiguousIds']}  starts at base+1: {t3['startsAtBasePlusOne']}")
    lines.append(f"- id range: {t3['idRange']}")
    lines.append(f"- texts missing from recall: {len(t3['missingFromRecall'])}")
    w = t3["writeMs"]
    lines.append(f"- write latency under load: p50={fmt_ms(w['p50'])}  p95={fmt_ms(w['p95'])}  p99={fmt_ms(w['p99'])}  max={fmt_ms(w['max'])}")
    lines.append("")
    lines.append("## Test 4 — latency (light load)")
    lines.append("")
    for name in ("remember", "recall", "e2e"):
        s = t4[name]
        lines.append(f"- **{name}**: mean={fmt_ms(s['mean'])}  p50={fmt_ms(s['p50'])}  p95={fmt_ms(s['p95'])}  p99={fmt_ms(s['p99'])}")
    lines.append("")
    lines.append("## Interpretation")
    lines.append("")
    lines.append(
        "Read-after-write is unconditional in this model: the DO-shaped authority commits "
        "the row inside the lock and only then returns a receipt, so any subsequent reader — "
        "the writer or another client — that hits the same authority sees it. There is no "
        "replication lag to hide, because there is exactly one place memory lives."
    )
    lines.append("")
    lines.append(
        "Concurrency holds because writes are serialized through `BEGIN IMMEDIATE` under a "
        "global threading lock — the local stand-in for a Durable Object's single-threaded "
        "execution. Ids are contiguous; ulids are unique; every payload is recallable."
    )
    lines.append("")
    lines.append("## Honest limitations of this prototype")
    lines.append("")
    lines.append(
        "- Local fsync is not a Cloudflare durability proof. WAL+`synchronous=NORMAL` is "
        "weaker than DO storage's distributed commit; numbers will differ on real CF."
    )
    lines.append(
        "- `LIKE` recall, not FTS5. Production Dejavu uses FTS5; this experiment intentionally "
        "tests authority/visibility, not ranking."
    )
    lines.append(
        "- No network: clients and the DO share a loopback socket. CF-side latency, DO "
        "placement, and cold-start cost are not modeled here."
    )
    lines.append(
        "- No auth, no quotas, no eviction, no embeddings — those are orthogonal."
    )
    lines.append(
        "- One DO id only. The shard/routing question (one DO per owner? per workspace?) "
        "is not in scope; this prototype models the inside of one authority."
    )
    lines.append("")
    return "\n".join(lines) + "\n"


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--base", required=True)
    parser.add_argument("--out", type=Path, required=True)
    args = parser.parse_args()
    run(args.base, args.out)


if __name__ == "__main__":
    main()
